lib/active_model/one_time_password.rb

```diff
@@ -28,30 +31,56 @@ module InstanceMethodsOnActivation
       def otp_regenerate_secret
         self.otp_column = ROTP::Base32.random_base32
       end
+      def otp_regenerate_counter
+        self.otp_counter = 0
+      end
 
       def authenticate_otp(code, options = {})
-        totp = ROTP::TOTP.new(self.otp_column, {digits: self.otp_digits})
-        if drift = options[:drift]
-          totp.verify_with_drift(code, drift)
+        if self.otp_counter_based
+          hotp = ROTP::HOTP.new(self.otp_column, {digits: self.otp_digits})
+          result = hotp.verify(code, self.otp_counter)
+          if result && options[:auto_increment]
+            self.otp_counter += 1
+            self.save if !self.new_record?
+          end
+          result
         else
-          totp.verify(code)
+          totp = ROTP::TOTP.new(self.otp_column, {digits: self.otp_digits})
+          if drift = options[:drift]
+            totp.verify_with_drift(code, drift)
+          else
+            totp.verify(code)
+          end
         end
       end
 
       def otp_code(options = {})
-        if options.is_a? Hash
-          time = options.fetch(:time, Time.now)
-          padding = options.fetch(:padding, true)
+        if self.otp_counter_based
+          if options[:auto_increment]
+            self.otp_counter += 1
+            self.save if !self.new_record?
+          end
+          ROTP::HOTP.new(self.otp_column, {digits: self.otp_digits}).at(self.otp_counter)
```

Comment:
Line is too long. [89/80]<br>Redundant curly braces around a hash parameter.<br>Redundant `self` detected.<br>Space inside { missing.<br>Space inside } missing.